LDAP group to local role mapping support

traceon opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category:
- New Feature

Changelog entry:
- Added support of mapping LDAP group names, and attribute values in general, to local roles for users from ldap user directories

Detailed description:
- Documented in a form of in-line comments in the sample `config.xml` file. See `<user_directories>` and `<ldap>` tags.
- Additionally, closes #18438.

traceon committed:
WIP: Implement group extraction and role mapping

traceon committed:
Compilation fixes

traceon committed:
Add regex syntax checks

traceon committed:
Change some config parameters to handle placeholders

traceon committed:
Actualize
Add role_mapping documentation (comments)

traceon committed:
Revert user directory name change

traceon committed:
Change naming
Add serialization of new params in getStorageParamsJSON()

traceon committed:
Improve regex instance creation code

traceon committed:
Merge branch 'master' into ldap-role-mapping

* master: (159 commits)
  Review fix.
  Update version_date.tsv after release 20.8.7.15
  wrong translation
  Update version_date.tsv after release 20.9.6.14
  Update version_date.tsv after release 20.10.5.10
  Update version_date.tsv after release 20.11.4.13
  Improvements in coverage images
  Fixed a problem with the translation of the document
  final_parallel
  final_parallel
  DOCSUP-4162: Document the system.replicated_fetches system table (#16900)
  Update settings.md
  Update settings.md
  Less verbose logging when fetch is impossible
  Don't add tons of client coverage files in stateful tests with coverage
  More compatible watches in TestKeeper
  Trying to make read_in_order_many_parts more stable
  trigger CI
  Update version_date.tsv after release 20.6.10.2
  Update visibleWidth.cpp
  ...

traceon committed:
Compilation fix

traceon committed:
Merge branch 'master' into ldap-role-mapping

* master: (50 commits)
  Update documentation-issue.md
  Add an option to use existing tables to perf.py
  DOCSUP-4280: Update the SELECT query (#17231)
  DOCSUP-3584 edit and translate (#17176)
  Fixed flaky test_storage_s3::test_custom_auth_headers
  Update 01560_merge_distributed_join.sql
  Minor improvements
  Slightly more correct
  Auto version update to [20.13.1.1] [54444]
  Auto version update to [20.12.1.5236] [54443]
  Update roadmap
  Add favicon; add loading indicator
  Fix race condition; history and sharing capabilities
  Update bitmap-functions.md
  Fix exception message
  Use default value for read-only flag in metadata for Disk3.
  ISSUES-16605 try fix review comment
  trigger CI
  ISSUES-16605 try fix integration failure
  ISSUES-16605 try fix integration test failure
  ...

vitlibar commented on programs/server/config.xml:
```diff
@@ -272,16 +271,67 @@
+                            Note, that the special characters must be escaped properly in XML.
+                        continue_on_match - flag that indicates, whether to continue matching and mapping using the subsequent rules
+                             after this rule successfully matched and mapped the string. By default, set to 'false'.
+                            If set to 'true' and multiple rules match, multiple role names may be generated from one same input string.
```
These rules look quite complex. Did you elaborate them by yourself or took from somewhere?

traceon replied:
No, this is not an adoption of something known. This is meant to be the most concise and generic way of transforming names, that covers any scenario we could imagine so far. The description may not be the best, but the essence of the rules is simple: chain of match+replace rules, where each of them produces a role name, if able to match the input string.

traceon replied:
Pushed a fix.

vitlibar replied:
It would better be an adoption ;)

Those regexp rules seem to be too difficult. And two extra boolean flags look you're trying to invent your own string processing programming language. And xml escaping definitely doesn't make the things easier.

Is it really necessary? I mean, just `search_filter`, `scope` and `attribute` are not enough? If not maybe one string-replacement rule can solve the problem. But if you're sure that we need a complex system with lots of replacement rules and control flow then I suppose it's better to enable something like AWK scripting.

traceon replied:
> It would better be an adoption ;)

Most of the systems don't have matching rules, they just assign names as-is. Clearly less flexible.

> Those regexp rules seem to be too difficult.

That's a generic regexp. Two fields, one for match and one for replace. It's hard for me to imagine anything simpler. :)
Implementing AWK would require the same number of config fields, but much more hustle in the code. And will be of same mental complexity as regexps, while less common.

> maybe one string-replacement rule can solve the problem.

Really, the same number of config fields, and much less flexible. And we may not solve the actual problems of our customers after all.

> Is it really necessary?

The thing is, there are many different environments, and we don't have all the requirements. So implementing something that will cover virtually all of the cases seems to be optimal way to spend resources.

traceon replied:
The fields have reasonable default values, so no need to specify all of them, to achieve typical behavior (of other systems).

vitlibar replied:
AWK is more general cause it cover almost everything which one can imagine about string processing. And it requires less fields in the config because instead of whole bunch of `<rule>`, `<match>`, `<replace>`, `<fail_if_all_rules_mismatch>`, `<continue_on_match>` you will have to specify only a path to *.awk file. I don't insist on using AWK though however it seems to be a decent alternative.

traceon replied:
It would be harder to integrate, and performance impact would be noticeable. But we can add it too, if the approach is proven to be more convenient.

vitlibar replied:
> harder to integrate

I agree, it would require an extra library in our `contrib` and it doesn't seem necessary right now.

vitlibar replied:
> That's a generic regexp. Two fields, one for match and one for replace. It's hard for me to imagine anything simpler. :)

For example, you cannot make string lowercased or uppercased with regexps. You cannot extract substring by indices. And you have troubles even writing those regexps with all that xml escaping.

If you like regexps, let's keep them. But I believe we should do something about those boolean flags, they're hard to explain and hard to understand. I suggest making `<apply_all_rules>` as a replacement for  `continue_on_match=true`, ..., `continue_on_match=true` and `<apply_any_rule>` as a replacement for `continue_on_match=false`, ..., `continue_on_match=false`. And `fail_if_all_rules_mismatch` doesn't seem to be very necessary at all: if no rules match then probably role will not be found and so user will get his error message anyway.

I just want this feature to be usable. That's why I think too much complexity is bad.

traceon replied:
I don't see how this is more simple. Same two flags, less flexibility. Regarding to-lower/to-upper, since we never heard about such use case, let's skip it until requested. Let's focus on a first iteration, then we will adjust it based on a feedback, not on our guesses.

By the way, it makes no sense for role names to be case sensitive.

vitlibar replied:
It's more about usability than programming. I feel we maybe should ask someone else. @filimonov can you say something about it?

filimonov replied:
I like simple things. Conventions over complicated configurations. Less flexible but 100% clear and predictable.

I would be happy with something like: if user will have a group with 'clickhouse_' prefix clickhouse will assign such a role (w/o prefix) to the user. Simple to explain, you don't need to look 15 mins in configurations to guess/understand what kind of privilegues the particular user will get after applying all those regexps. 

I have no clear understanding what advantages those complicated rules give (and I'm bit scared of how i will need to explain that to somebody) 

Maybe we can go with simple setup with predefined / simple matching rule at the beginning, and start extending it when smbd will came with  such a requirement

traceon replied:
Changed to prefix.

vitlibar commented on programs/server/config.xml:
```diff
@@ -272,16 +271,67 @@
+                         applied.
+                    base_dn - template used to construct the base DN for the LDAP search.
+                            The resulting DN will be constructed by replacing all '{username}' and '{bind_dn}' substrings
+                             of the template with the actual user name and bind DN during each LDAP search.
```
Where can `base_dn` be used? If the only place is `search_filter` then it seems we don't need `base_dn` cause everything can be put directly to `search_filter`.

traceon replied:
These is very LDAP-specific params, the base DN is specified in ldap_search routine as one of the parameters, the filter is another one. The base DN is a path where the search should start from. Changing their places or names would be very misleading for LDAP admins.

traceon replied:
As a substitution, `base_dn` may be used in the filter though. But that depend on the query.

traceon committed:
Merge branch 'master' into ldap-role-mapping

* master: (207 commits)
  Update RadixSort.h
  rerun tests to be sure
  Update date_time_short perf test for toUnixTimestamp(Date())
  update test
  remove comments
  better
  fix tests
  style
  update copy pasted test
  better
  comments
  better merge
  new interface for the function
  better
  Fix comments
  Add missing file
  Make the code less bad
  initial
  test added
  style
  ...

vzakaznikov committed:
Updating TestFlows tests to support changes for LDAP role mapping.

vitlibar commented on src/Access/Authentication.h:
```diff
@@ -90,6 +96,7 @@ class Authentication
     /// Checks if the provided password is correct. Returns false if not.
     /// User name and external authenticators' info are used only by some specific authentication type (e.g., LDAP_SERVER).
     bool isCorrectPassword(const String & password_, const String & user_, const ExternalAuthenticators & external_authenticators) const;
+    bool isCorrectPasswordLDAP(const String & password_, const String & user_, const ExternalAuthenticators & external_authenticators, const LDAPSearchParamsList * search_params = nullptr, LDAPSearchResultsList * search_results = nullptr) const;
```
It doesn't look good to me. We had a single function `isCorrectPassword()` and now we're going to have an extra function just for one authentication type. It seems you could easily avoid that because there is one place where those extra arguments are passed from - it's `LDAPAccessStorage`, and `LDAPAccessStorage` always uses only the same authentication type and `LDAPAccessStorage` could call `ldap_client.authenticate()` directly without spoiling the general interface.

traceon replied:
I didn't want to implement the authentication logic in different places. That would make the code misleading and hard to maintain.

traceon replied:
By the way, the idea of `Credential` would solve this particular problem quite nicely: the authentication logic would be encapsulated in each `Credential` implementation, and accessed by a common interface inside `Authentication` class calls.

traceon replied:
> LDAPAccessStorage could call ldap_client.authenticate() directly without spoiling the general interface.

Once #15988 is merged, I'll be able to call `ExternalAuthenticators::checkLDAPCredentials(...)` from `LDAPAccessStorage` directly...

vitlibar replied:
https://github.com/ClickHouse/ClickHouse/pull/15988 merged

traceon replied:
Changed.

vitlibar reviewed (changes requested):
https://github.com/ClickHouse/ClickHouse/pull/17211#discussion_r543338650

filimonov commented on programs/server/config.xml:
```diff
@@ -300,16 +299,67 @@
+                        <search_filter>(&amp;(objectClass=groupOfNames)(member={bind_dn}))</search_filter>
+                        <fail_if_all_rules_mismatch>true</fail_if_all_rules_mismatch>
+                        <rule>
+                            <match>clickhouse_(.+)</match>
```
Imagine `i_want_to_hack_clickhouse_admin` group... 

Put anchors! If you did that mistake the users will do even worse mistakes...

filimonov commented on programs/server/config.xml:
```diff
@@ -300,16 +299,67 @@
                     <server>my_ldap_server</server>
                     <roles>
                         <my_local_role1 />
                         <my_local_role2 />
```
Whitelist of the acceptable roles?

It's not handy to edit xml file every time when new role should be added... Lives separately from roles itself. What if listed role was dropped / never existed?

filimonov commented on programs/server/config.xml:
```diff
@@ -300,16 +299,67 @@
                         <my_local_role1 />
                         <my_local_role2 />
                     </roles>
+                    <role_mapping>
```
can it have SEVERAL `<role_mapping>` sections?

traceon replied:
Yes.

filimonov commented on programs/server/config.xml:
```diff
@@ -300,16 +299,67 @@
+                             substrings of the template with the actual user name, bind DN, and base DN during each LDAP search.
+                            Note, that the special characters must be escaped properly in XML.
+                    fail_if_all_rules_mismatch - flag to trigger failure if none of the rules were able to match and transform any
+                             of the resulting strings returned by the LDAP search. By default, set to 'true'.
```
What is alternative to triggering failure? Allow user to login w/o any roles assigned?

traceon replied:
Not relevant anymore.

traceon committed:
Do transformations based on prefix only

traceon committed:
Simplify the code

vzakaznikov committed:
Merge branch 'master' of https://github.com/traceon/ClickHouse into ldap-role-mapping

vzakaznikov committed:
Updating ldap/external_user_directory/requirements/requirements.py

traceon committed:
Revisit mapped role management

vitlibar commented on programs/server/config.xml:
```diff
@@ -371,16 +370,40 @@
+                            Note, that the special characters must be escaped properly in XML.
+                    prefix - prefix, that will be expected to be in front of each string in the original list of strings returned by
+                             the LDAP search. Prefix will be removed from the original strings and resulting strings will be treated
+                             as local role names. Empty, by default.
```
I believe this explanation can be made clearer. Instead of the current text 

`The list of strings returned by the search will be transformed...`

which is a bit vague I think we could be more precise and write something like

`when a user logins there is a search occurs in LDAP using search_filter
and the name of the logged user, and for each found entry a specified
attribute extracted which after removing a specified prefix becomes name
of a role inside ClickHouse which should be created beforehand by
CREATE ROLE command.`

traceon replied:
Changed.

vitlibar commented on programs/server/config.xml:
```diff
@@ -371,16 +370,40 @@
-                         will fail as if the provided password was incorrect.
+                        If no roles are specified here or assigned during role mapping (below), user will not be able to perform any
+                         actions after authentication.
+                role_mapping - section with LDAP search parameters and mapping rules.
```
Maybe it's better to rename `role_mapping` into `roles_search`? Just to represent the fact that all those parameters are about searching in LDAP and to make the difference between this section and the `roles` section more clear.

traceon replied:
`imported_roles` maybe?

vitlibar commented on programs/server/config.xml:
```diff
@@ -371,16 +370,40 @@
+                            The resulting filter will be constructed by replacing all '{username}', '{bind_dn}', and '{base_dn}'
+                             substrings of the template with the actual user name, bind DN, and base DN during each LDAP search.
+                            Note, that the special characters must be escaped properly in XML.
+                    prefix - prefix, that will be expected to be in front of each string in the original list of strings returned by
```
What if the attribute doesn't have a specified prefix in its value? It won't be used as a role name then? I'm not sure but maybe it's better to rename `prefix` into `remove_prefix` to emphasize the fact we remove the prefix (and we don't insert the prefix).

traceon replied:
As we learned from the feedback from people who routinely configure such things for enterprise, it's a common (and effectively the only) way of mapping the role names, so I think it's better to keep it `prefix` since the underlying functionality is quite familiar to everyone.

vitlibar commented on programs/server/config.xml:
```diff
@@ -371,16 +370,40 @@
+                    base_dn - template used to construct the base DN for the LDAP search.
+                            The resulting DN will be constructed by replacing all '{username}' and '{bind_dn}' substrings
+                             of the template with the actual user name and bind DN during each LDAP search.
+                    attribute - attribute name whose values will be returned by the LDAP search.
```
Maybe it's better to describe `attribute` after `search_filter` (but before `prefix`) to represent the order of how all these parameters are used.

traceon replied:
Changed.

vitlibar commented on programs/server/config.xml:
```diff
@@ -371,16 +370,40 @@
+                    scope - scope of the LDAP search.
+                            Accepted values are: 'base', 'one_level', 'children', 'subtree' (the default).
+                    search_filter - template used to construct the search filter for the LDAP search.
+                            The resulting filter will be constructed by replacing all '{username}', '{bind_dn}', and '{base_dn}'
```
I think `{user_name}` looks better.

traceon replied:
Changed.

vitlibar commented on programs/server/config.xml:
```diff
@@ -371,16 +370,40 @@
+                    scope - scope of the LDAP search.
+                            Accepted values are: 'base', 'one_level', 'children', 'subtree' (the default).
+                    search_filter - template used to construct the search filter for the LDAP search.
+                            The resulting filter will be constructed by replacing all '{username}', '{bind_dn}', and '{base_dn}'
```
About the replacement of `{bind_dn}` and `{base_dn}` in `search_filter`, they look quite unnecessary and a bit misleading. Are you sure there is a real case about them?

traceon replied:
Yes, some use cases may require search filter to contain those.

traceon committed:
Merge branch 'master' into ldap-role-mapping

* master: (620 commits)
  Add test for some possible ambiguities in syntax
  Update PushingToViewsBlockOutputStream.h
  [For #18707] MySQL compatibility: support DIV and MOD operators
  Mark another flaky test
  Remove some headers
  Mark some TestFlows as flaky
  Fix error
  Fix errors
  One more test
  Arcadia does not support distributed queries
  Add a test for #14974
  Added a test from #15641
  More robust stateful test
  Update tests
  Remove bad code in HashJoin
  Update test
  Don't allow conversion between UUID and numeric types
  Remove pink screen with confusing questions about Kerberos
  Do not throw from Parser
  Fix the unexpected behaviour of show tables when antlr parser enabled (#18431)
  ...

# Conflicts:
#	programs/server/config.xml
#	src/Access/Authentication.cpp
#	src/Access/Authentication.h

traceon committed:
Rename {username} to {user_name}
Add caching/checking of search_params
Adjust comments/doc
Use special authentication logic from ExternalAuthenticators::checkLDAPCredentials

traceon committed:
Merge branch 'master' into ldap-role-mapping

* master: (605 commits)
  DOCSUP-4710: Added support numeric parameters in number and string data types (#18696)
  DOCSUP-5604: Edit and translate to Russian (#18929)
  Update version_date.tsv after release 21.1.2.15
  Usability improvement of clickhouse-test
  Update jit_large_requests.xml
  Update README.md
  Update images.json
  Make symbolizers available in fuzzer Docker image
  Update Dragonbox
  Speed up aggregate function sum
  Fix MSan report in Kerberos library
  Fix MSan error in rocksdb #19213
  Add more Fuzzer tasks
  Fixes
  Update comment for curl dependency for aws
  Disable curl for mariadb-connector-c (it is not required)
  Fix TSan
  Skip test for ANTLR
  DistributedBlockOutputStream: add more comments
  DistributedBlockOutputStream: Remove superfluous brackets for string construction
  ...

vzakaznikov committed:
Enabling all TestFlows modules.
Increasing clickhouse container health check timeouts.

vitlibar commented:
The problem with AST fuzzer has been already fixed (https://github.com/ClickHouse/ClickHouse/issues/19287)

olgarev commented:
@vitlibar Hi! It seems that this functionality was already documented here https://github.com/ClickHouse/ClickHouse/pull/20208. Am I right?

traceon commented:
@olgarev Yes, that's correct.
